### DISCUSSION TITLE
[Noob alert] Image border blurring

### DISCUSSION THREAD

**Discussion by WingHaa** [ImageMagick/ImageMagick#7703] [Help] [answered]
### ImageMagick version

7.1.1-39

### Operating system, version and so on

unix Archlinux 6.6.57-1

### Description

Hello! I'm very new to image editting in general and image magick specifically. 

I have tried researching for the past week but got no fruit. I have tried Fred's border effect script but it doesn't come very close and I'm afraid I'm too unfamiliar with this sort of stuff to try tweaking it.
I would be very appreciate if anyone can help produce or point me the direction to get the following effect. The background of output image is supposed to be transparent so it can blend to the brown background.
- Source image:
![05-baby](https://github.com/user-attachments/assets/395091c2-250e-4356-90f9-74ba28a697b9)

- Desired output (background need to be transparent):
![mpv-shot0003](https://github.com/user-attachments/assets/acbd01d6-816f-4063-a0ce-ad1a4ffa4f95)

**Comment by snibgo** [comment]
Your question is vague. Is the border effect merely a change in transparency, or is there also a change in colour?

Your "desired output" is jpg, which cannot record transparency.

A simple border could be (bash syntax):
```
magick.exe \
  babyToes.jpg \
  \( +clone \
     -fill White -colorize 100 \
     -virtual-pixel Black \
     -blur 0x40 \
     -auto-level \
  \) \
  -alpha off -compose CopyOpacity -composite \
  x.webp
```

**Comment by WingHaa** [comment]
I'm sorry for the poor explanation but I find it pretty hard to describe since I'm not really familiar with image processing...
I don't have the original output so I pasted the jpg as output.

Basically the four border of the image (red rectangle) have this "pixelated" or "trail" effect for the lack of better word. These effect have the original color while the brown part (blue rectangle) are supposed to be transparent. The brown come from an underneath brown background image which was put under the transparent image in photoshop or premiere.
![image](https://github.com/user-attachments/assets/c6625f4b-d24f-43b9-91b8-919ca3a40e0d)

**Comment by snibgo** [comment]
Does my command do what you want? If not, what is wrong with the result?

**Comment by fmw42** [comment, marked as answer]
Perhaps this is what you want?

```
magick \
babyToes.jpg \
\( +clone \
-fill "gray(100%)" \
-colorize 100 \
-virtual-pixel Black \
-blur 0x40 \
-auto-level \
-attenuate 0.75 \
+noise gaussian \
\) \
-compose multiply \
-composite \
x.webp
```

[x.webp.zip](https://github.com/user-attachments/files/17439775/x.webp.zip)

Here is a PNG output:

![x](https://github.com/user-attachments/assets/8c29baa5-64a9-4110-82c3-1871e8ad066e)

**Comment by WingHaa** [comment]
Thank you Fred. That was what I wanted.
Also thank snibgo for trying to help despite my vagueness

**Comment by WingHaa** [comment]
@fmw42 
I tested on other images and noticed that the noise is applied to the entire image. Is there anyway to make it not apply to inner red rectangle here?
![out](https://github.com/user-attachments/assets/12857a2c-9447-4064-959b-c543b7fa79a1)

**Reply by fmw42** [reply]
Where does the red border come from?  Will that always be used? If so, I can limit the noise to the area outside the red border.

**Reply by WingHaa** [reply]
Oh. In the process there is no red border. I was using it to better describe how I want it. Is it not possible without the border?

**Reply by fmw42** [reply]
Try this

```
magick \
babyToes.jpg \
\( \
+clone \
-fill white \
-colorize 100 \
-virtual-pixel Black \
-blur 0x40 \
-auto-level \
\( \
+clone \
-threshold 95% \
-blur 0x40 \
-negate \
-write mpr:mask \
+delete \
\) \
-seed 123 \
-attenuate 0.75 \
+noise gaussian \
\) \
mpr:mask -compose multiply \
-composite \
x2.png

```

or (restructured):

```
magick \
babyToes.jpg \
\( +clone -fill white -colorize 100 -virtual-pixel Black -blur 0x40 -auto-level \
\( +clone -threshold 95% -blur 0x40 -negate -write mpr:mask +delete \) \
-seed 123 -attenuate 0.75 +noise gaussian \) \
mpr:mask -compose multiply -composite \
x2.png
```


![x2](https://github.com/user-attachments/assets/2014f278-1cf3-4a10-a8b8-f7ce332e781c)

**Reply by fmw42** [reply]
Is that what you want?

**Reply by WingHaa** [reply]
It's almost there. Is it possible to make the black noise transparent (converting them to alpha channel) ?
![image](https://github.com/user-attachments/assets/d4e163a1-0c79-427c-a8e8-8ab364dacd17)

**Reply by fmw42** [reply]
I don't see how that would be any different in the resulting output. You are already seeing the underlying image through the noise. 

Why do you want that?  Do you need to see through the image to some background image or just see through the noise to your your image. Post the image above before adding the noise border so I can test with it.  

What tool have you been using to create these images. It might help if I know what is being used and how and the arguments.

**Reply by WingHaa** [reply]
My purpose is to use these images to combine with some background using ffmpeg with a bunch of complex filters to make a video out of them. So I would need to be able to see the underlying background through the noises. Currently I'm seeing black noises instead of the background.

Here's the overlay command:
```
 ffmpeg -i input.png -i image.png -filter_complex "[0:v][1:v] overlay=(main_w-overlay_w)/2:(main_h-overlay_h)/2" output.png
 ```

Here's the image without the noise:
 ![input](https://github.com/user-attachments/assets/27da2ed7-5f9e-4127-aa38-da183edb1df6)

**Reply by fmw42** [reply]
So your want the result to be transparent only where there is noise and to the degree that the noise is dark. So you want the noise to fade to pure transparent at the edges of the results where your example above shows fading to black. I do not see your result above showing transparent below the noise in the output.

Or are you just asking for the result to be padded with transparency as in your example (i.e. only outside the resulting image).

Your request and your example do not match.

I will look into this tomorrow or Monday as I have time.  Please respond to my comments/questions above in this reply.

**Reply by fmw42** [reply]
Try this

```
magick \
momBaby.jpg \
\( +clone -fill white -colorize 100 -virtual-pixel Black -blur 0x40 -auto-level \
-threshold 75% -blur 0x40 -seed 123 -attenuate 0.75 +noise gaussian \) \
-alpha off -compose copy_opacity -composite \
x3.png
```

**Reply by WingHaa** [reply]
> Your request and your example do not match.

I'm really sorry as it seems I have lead you on a wild goose chase. I will try to show more by examples.

With your above command, when I tried to apply it to the video the main part got really pronounce pixelated.
I have also tried fiddling with blur parameter in the second pic and I think it come close to the sample. But it eat too much image and you can still see the rectangle around the image.

Original command:
![mpv-shot0002](https://github.com/user-attachments/assets/68194b3f-0e92-4d01-bb0f-bbb55c642cc6)

With a higher blur value:
![mpv-shot0001](https://github.com/user-attachments/assets/96ec867b-d587-44e0-a587-bdd1fda9b2b6)

I want a image that when overlay with a back ground that will come out like these (please ignore the decorations). Here we are using a brown background:
![image](https://github.com/user-attachments/assets/a3c1cb95-897c-420d-b249-cbff028bc6f0)
![image](https://github.com/user-attachments/assets/863e9250-07af-45c1-9cf6-e35e8ff88918)

**Reply by fmw42** [reply]
provide me the background image so that I can see how the border looks on top of it.

**Reply by lmquansgt** [reply]
Here's the background:
![b4608e1e91f96ce34020c18bcb09950f](https://github.com/user-attachments/assets/a9a32544-8841-415a-8867-7f3ac40dff82)

**Reply by fmw42** [reply]
This image is smaller than the other two.  How are you resizing the others to fit?  You never mentioned the need to resize!

What do you want to do here with respect to resizing the foreground images?

You keep changing the requirements! Every time I send you something you change the requirements.  This is not just answering a question or providing some simple code. This forum is not a code writing service. People help as a curtesy but you should be clear about what you require from the beginning. (Or you can ask on the Consulting forum if you are willing to pay for help). 

I will try to help, but not until at least tomorrow. It is late for me now.

**Reply by WingHaa** [reply]
I didn't realized I sent the smaller version of the bg. The resizing doesn't matter. I just need to overlay the output on the background.

I think the problem is I don't have enough knowledge about this as a whole to be clear about my requirements. I'm very appreciated your help so far. Honestly no holds barred if you decided to stop helping. 
It's just that I can't afford paid consulting for this issue...

Here's the bigger version of the background.

![bg](https://github.com/user-attachments/assets/9b36da7d-d871-45c8-8c81-24b0c13b7b5b)

**Reply by fmw42** [reply]
This version is much larger than the other two images. So either way, a resize is necessary. 

So do you want to a) process the original foreground to add the border and then resize and overlay OR b) resize first then process to add the border, then overlay.

Also do you want me to resize the foreground smaller to fit the smaller background or resize it larger to fit the larger background?

**Reply by WingHaa** [reply]
In that case please go with a) and resize it smaller to fit the smaller background. 
Thank you for your help.

**Reply by fmw42** [reply]
If I do a) then the noise will be smaller and more compact and may not show as well. Should I go ahead with that?

**Reply by WingHaa** [reply]
I didn't think of that. The noise need to be visible so if b) don't have problem with that please do b) instead.

**Comment by fmw42** [comment]
Here is another approach:

```
magick \
momBaby.jpg -write mpr:img -set option:ww "%w" -set option:hh "%h" \
\( -clone 0 -fill gray -colorize 100 -seed 123 -attenuate 2 +noise random \
-channel g -separate -white-threshold 50% -write mpr:noise  +delete \) \
\( -clone 0 -fill white -colorize 100 -shave 1 -bordercolor black -border 1 -virtual-pixel Black \
-gamma 2 +level 0,100 -white-threshold 99  \
-morphology Distance Euclidean:4,50! +write t2.png -write mpr:ramp +delete \) \
\( mpr:noise mpr:ramp -fx "v<1?u*v:white"  \) \
-alpha off -compose copy_opacity -composite  \
\( background.jpeg -resize "%[fx:ww+100]x%[fx:hh+100]!" \) +swap \
-gravity center -compose over -composite \
x4.png
```

![x4](https://github.com/user-attachments/assets/bbb6ebbe-4205-42db-973c-64a5c39b5ff5)
